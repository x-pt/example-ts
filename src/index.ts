import * as fs from "node:fs/promises";
import * as core from "@actions/core";
import axios from "axios";
import * as toml from "toml";

interface Config {
    input_text?: string;
    find_word?: string;
    replace_word?: string;
    number_list?: number[];
    input_file?: string;
    output_file?: string;
    append_text?: string;
    api_url?: string;
}

async function checkAPIReachability(apiUrl: string): Promise<void> {
    try {
        const response = await axios.get(apiUrl, { timeout: 10000 });
        if (response.status < 200 || response.status >= 300) {
            core.warning(`API is not reachable, status code: ${response.status}`);
        } else {
            core.info(`API ${apiUrl} is reachable.`);
        }
    } catch (error) {
        core.warning(`Failed to make API request: ${error instanceof Error ? error.message : String(error)}`);
    }
}

async function readAndAppendToFile(inputFile: string, outputFile: string, appendText: string): Promise<void> {
    try {
        const content = await fs.readFile(inputFile, "utf-8");
        const modifiedContent = `${content}\n${appendText}`;
        await fs.writeFile(outputFile, modifiedContent, { encoding: "utf-8" });
        core.info(`Appended text to file: ${outputFile}`);
    } catch (error) {
        throw new Error(`File operation failed: ${error instanceof Error ? error.message : String(error)}`);
    }
}

function processText(
    text: string,
    findWord: string,
    replaceWord: string,
): {
    processedText: string;
    wordCount: number;
} {
    const processedText = text.replace(new RegExp(findWord, "g"), replaceWord);
    const wordCount = processedText.trim() === "" ? 0 : processedText.trim().split(/\s+/).length;
    return { processedText, wordCount };
}

function calculateNumberStats(numbers: number[]): { sum: number; average: number } {
    const sum = numbers.reduce((acc, num) => acc + num, 0);
    const average = numbers.length > 0 ? sum / numbers.length : 0;
    return { sum, average };
}

export async function run(): Promise<void> {
    try {
        const configPath = core.getInput("config_path") || ".github/configs/setup-custom-action-by-ts.toml";
        const configContent = await fs.readFile(configPath, "utf-8");
        const config: Config = toml.parse(configContent);

        const {
            input_text = "",
            find_word = "",
            replace_word = "",
            number_list = [],
            input_file = "",
            output_file = "",
            append_text = "",
            api_url = "",
        } = config;

        if (api_url) {
            await checkAPIReachability(api_url);
        }

        if (input_file && output_file && append_text) {
            await readAndAppendToFile(input_file, output_file, append_text);
        }

        const { processedText, wordCount } = processText(input_text, find_word, replace_word);
        const { sum, average } = calculateNumberStats(number_list);

        core.setOutput("processed_text", processedText);
        core.setOutput("word_count", wordCount);
        core.setOutput("sum", sum);
        core.setOutput("average", average);
    } catch (error) {
        core.setFailed(`Action failed with error: ${error instanceof Error ? error.message : String(error)}`);
    }
}

if (!process.env.JEST_WORKER_ID) {
    run().catch((error) => {
        console.error("Unhandled error:", error);
        process.exit(1);
    });
}
