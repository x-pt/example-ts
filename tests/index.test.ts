import * as fs from "node:fs/promises";
import * as core from "@actions/core";
import axios from "axios";
import { run } from "../src";

jest.mock("@actions/core");
jest.mock("node:fs/promises");
jest.mock("axios");

const mockConfig = `
input_text = "Hello world! Hello!"
find_word = "Hello"
replace_word = "Hi"
number_list = [1, 2, 3, 4, 5]
input_file = "input.txt"
output_file = "output.txt"
append_text = "Goodbye!"
api_url = "https://api.example.com/data"
`;

describe("GitHub Action", () => {
    const mockGetInput = jest.spyOn(core, "getInput");
    const mockSetOutput = jest.spyOn(core, "setOutput");
    const mockSetFailed = jest.spyOn(core, "setFailed");
    const mockWarning = jest.spyOn(core, "warning");
    const mockInfo = jest.spyOn(core, "info");
    const mockReadFile = fs.readFile as jest.MockedFunction<typeof fs.readFile>;
    const mockWriteFile = fs.writeFile as jest.MockedFunction<typeof fs.writeFile>;
    const mockAxiosGet = axios.get as jest.MockedFunction<typeof axios.get>;

    beforeAll(() => {
        mockGetInput.mockReturnValue(".github/configs/setup-custom-action-by-ts.toml");
    });

    beforeEach(() => {
        jest.resetAllMocks();
        setupDefaultMocks();
    });

    const setupDefaultMocks = () => {
        mockReadFile.mockResolvedValue(mockConfig);
        mockAxiosGet.mockResolvedValue({ status: 200, data: {} });
    };

    const setupFileMocks = (inputContent: string) => {
        mockReadFile.mockImplementation((path) => {
            if (path === ".github/configs/setup-custom-action-by-ts.toml") {
                return Promise.resolve(mockConfig);
            }
            if (path === "input.txt") {
                return Promise.resolve(inputContent);
            }
            return Promise.reject(new Error("Unexpected file read"));
        });
    };

    describe("Text Processing", () => {
        it("should process text, count words, calculate sum and average correctly", async () => {
            await run();

            expect(mockSetOutput.mock.calls).toMatchSnapshot();
        });

        it("should handle empty input text", async () => {
            mockReadFile.mockResolvedValue(`
input_text = ""
number_list = []
`);

            await run();

            expect(mockSetOutput.mock.calls).toMatchSnapshot();
        });
    });

    describe("File Operations", () => {
        it("should read from the input file and append text to the output file correctly", async () => {
            setupFileMocks("Original file content.");

            await run();

            expect(mockReadFile).toHaveBeenCalledWith("input.txt", "utf-8");
            expect(mockWriteFile).toHaveBeenCalledWith("output.txt", "Original file content.\nGoodbye!", {
                encoding: "utf-8",
            });
            expect(mockInfo).toHaveBeenCalledWith("Appended text to file: output.txt");
        });

        it("should handle missing input file gracefully", async () => {
            mockReadFile.mockRejectedValue(new Error("File not found"));

            await run();

            expect(mockSetFailed).toHaveBeenCalledWith("Action failed with error: File not found");
        });

        it("should handle file write errors gracefully", async () => {
            setupFileMocks("Original file content.");
            mockWriteFile.mockRejectedValue(new Error("Failed to write to output file"));

            await run();

            expect(mockSetFailed).toHaveBeenCalledWith(
                "Action failed with error: File operation failed: Failed to write to output file",
            );
        });
    });

    describe("API Operations", () => {
        it("should check API reachability correctly", async () => {
            await run();

            expect(mockAxiosGet).toHaveBeenCalledWith("https://api.example.com/data", { timeout: 10000 });
            expect(mockInfo).toHaveBeenCalledWith("API https://api.example.com/data is reachable.");
        });

        it("should handle API request failures gracefully", async () => {
            mockAxiosGet.mockRejectedValue(new Error("API error"));

            await run();

            expect(mockWarning).toHaveBeenCalledWith("Failed to make API request: API error");
        });

        it("should warn on bad API status", async () => {
            mockAxiosGet.mockResolvedValue({ status: 500, data: {} });

            await run();

            expect(mockWarning).toHaveBeenCalledWith("API is not reachable, status code: 500");
        });
    });

    describe("Configuration Handling", () => {
        it("should use default values when config is empty", async () => {
            mockReadFile.mockResolvedValue("");

            await run();

            expect(mockSetOutput.mock.calls).toMatchSnapshot();
        });

        it("should handle missing configuration file gracefully", async () => {
            mockReadFile.mockRejectedValue(new Error("Config file not found"));

            await run();

            expect(mockSetFailed).toHaveBeenCalledWith("Action failed with error: Config file not found");
        });
    });
});
