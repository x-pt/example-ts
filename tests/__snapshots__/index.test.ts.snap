// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitHub Action Configuration Handling should use default values when config is empty 1`] = `
[
  [
    "processed_text",
    "",
  ],
  [
    "word_count",
    0,
  ],
  [
    "sum",
    0,
  ],
  [
    "average",
    0,
  ],
]
`;

exports[`GitHub Action Text Processing should handle empty input text 1`] = `
[
  [
    "processed_text",
    "",
  ],
  [
    "word_count",
    0,
  ],
  [
    "sum",
    0,
  ],
  [
    "average",
    0,
  ],
]
`;

exports[`GitHub Action Text Processing should process text, count words, calculate sum and average correctly 1`] = `
[
  [
    "processed_text",
    "Hi world! Hi!",
  ],
  [
    "word_count",
    3,
  ],
  [
    "sum",
    15,
  ],
  [
    "average",
    3,
  ],
]
`;
